contracts/0.8.9/WithdrawalQueue.sol

```diff
@@ -0,0 +1,255 @@
+     */
+    function finalize(
+        uint256 _lastIdToFinalize, 
+        uint256 _etherToLock,
+        uint256 _totalPooledEther,
+        uint256 _totalShares
+    ) external payable onlyOwner {
+        require(
+            _lastIdToFinalize >= finalizedQueueLength && _lastIdToFinalize < queue.length, 
+            "INVALID_FINALIZATION_ID"
+        );
+        require(lockedEtherAmount + _etherToLock <= address(this).balance, "NOT_ENOUGH_ETHER");
+
+        _updatePriceHistory(_totalPooledEther, _totalShares, _lastIdToFinalize);
+
+        lockedEtherAmount += _etherToLock;
+        finalizedQueueLength = _lastIdToFinalize + 1; 
+    }
+
+    /**
+     * @notice Mark `_requestId` request as claimed and transfer reserved ether to recipient
+     * @param _requestId request id to claim
+     * @param _priceIndexHint price index found offchain that should be used for claiming 
+     */
+    function claim(uint256 _requestId, uint256 _priceIndexHint) external returns (address recipient) {
+        // request must be finalized
+        require(finalizedQueueLength > _requestId, "REQUEST_NOT_FINALIZED");
+
+        Request storage request = queue[_requestId];
+        require(!request.claimed, "REQUEST_ALREADY_CLAIMED");
+
+        request.claimed = true;
+
+        Price memory price;
+
+        if (_isPriceHintValid(_requestId, _priceIndexHint)) {
+            price = finalizationPrices[_priceIndexHint];
+        } else {
+            // unbounded loop branch. Can fail
+            price = finalizationPrices[findPriceHint(_requestId)];
+        }
+
+        (uint256 etherToTransfer,) = _calculateDiscountedBatch(
+            _requestId, 
+            _requestId, 
+            price.totalPooledEther, 
+            price.totalShares
+            );
+        lockedEtherAmount -= etherToTransfer;
+
+        request.recipient.sendValue(etherToTransfer);
+    }
+
+    /**
+     * @notice calculates the params to fullfill the next batch of requests in queue
+     * @param _lastIdToFinalize last id in the queue to finalize upon 
+     * @param _totalPooledEther share price compoinent to finalize requests
+     * @param _totalShares share price compoinent to finalize requests
+     * 
+     * @return etherToLock amount of eth required to finalize the batch
+     * @return sharesToBurn amount of shares that should be burned on finalization
+     */
+    function calculateFinalizationParams(
+        uint256 _lastIdToFinalize,
+        uint256 _totalPooledEther,
+        uint256 _totalShares
+    ) external view returns (uint256 etherToLock, uint256 sharesToBurn) {
+        return _calculateDiscountedBatch(finalizedQueueLength, _lastIdToFinalize, _totalPooledEther, _totalShares);
+    }
+
+    function findPriceHint(uint256 _requestId) public view returns (uint256 hint) {
+        require(_requestId < finalizedQueueLength, "PRICE_NOT_FOUND");
+
+        for (uint256 i = finalizationPrices.length - 1; i >= 0; i--) {
```

Comment:
can't do this

if `i == 0`, then `i--` would lead to revert cause `i` is unsigned.